### ISSUE TITLE
Suggest to adjust C_INCLUDE_PATH and/or LIBRARY_PATH on package install failure

### ISSUE THREAD

**Issue by dbuenzli** [ocaml/opam#3049]
Whenever a package fails and has depexts, `opam install` suggests to run `opam depext`. 

It would be nice to add that if this was already done and successful, then the environment variables `C_INCLUDE_PATH` and/or `LIBRARY_PATH` might need to be adjusted if the software gets installed in a non standard location.

This was suggested by @ygrek [here](https://github.com/ocaml/opam-repository/issues/3093#issuecomment-312929998).

**Comment by davehayes** [comment]
Hello. It would be extremely helpful to document that these two variables exist and exactly what they do.

As someone trying to build the entire suite of ocaml package dependencies for liquidsoap on FreeBSD,  2/3 of the issues I encountered (so far) trying to do this were handled by pkg-config, but the other 1/3 ... well I was trying all sorts of hacks until I read this issue. With those two variables properly set, I have no issues and building things like lame (which do not have pkg-config) work.

Thank you for any effort to provide complete documentation!

**Comment by hcarty** [comment]
@davehayes These variables are used by the underlying C compiler and linker. It's probably worth mentioning them in the opam docs/FAQ somewhere but they are not specific to opam/OCaml.

**Comment by davehayes** [comment]
Ok thanks. It' s still worth mentioning them IMO, but I didn't know if opam isolated environment variables or had redefined old ones or what.

**Comment by dbuenzli** [comment]
@davehayes you can read about them [here](https://gcc.gnu.org/onlinedocs/gcc/Environment-Variables.html).

**Comment by davehayes** [comment]
@dbuenzli  FreeBSD 10 uses clang (3.4.1) now, not gcc. I build opam from ports, and in looking at the Makefile for opam there is no dependency on gcc that I see at first glance. Thus, while some of these variables actually work, there was no reason for me to assume that they would.

**Comment by dbuenzli** [comment]
@davehayes indeed I looked up the clang doc and `LIBRARY_PATH` doesn't seem to be mentioned. It seems `C_INCLUDE_PATH` is [however](https://www.freebsd.org/cgi/man.cgi?query=clang&sektion=1&manpath=FreeBSD+9.0-RELEASE#end), maybe you should report this upstream.

**Comment by davehayes** [comment]
That would be a good idea, but after some introspection it would appear that `configure` picked up these two environment variables and somehow "made it right".  The original failure I had was in the `configure` step.

**Comment by dbuenzli** [comment]
@davehayes it's a little bit unclear which `configure` you are talking about. Besides these environment are not supposed to be picked up by `configure` but by the underlying tools themselves.

**Comment by davehayes** [comment]
The original issue I had (which wasn't the only one) was that `opam install lame` ran a configure step as part of it's install process.  Setting these environment variables allowed lame to install. I believe the configure step that ran was the configure included in the bindings that were installed.

**Comment by AltGr** [comment]
While I see the point, just setting the variables before re-running opam looks like a slippery slope to me:
- it's a transient user state, so whenever opam tries to recompile the package, e.g. due to an updated dependency, it'll break again
- we may choose to run opam commands with a cleaned up environment sometime in the future

`opam pin edit <package>` then adding the appropriate `build-env: ` field would work, but that's probably too involved. Environment variables can also be set at the switch level, so we could provide `opam config setenv` for this purpose. The variable would then be re-exported by any `eval $(opam [config] env)`

**Comment by dbuenzli** [comment]
> While I see the point, just setting the variables before re-running opam looks like a slippery slope to me:

It doesn't really feel more slippery than now. opam and the tools it invokes has always been sensitive to whatever environment you run `opam` in. In any case the goal here is not to set the variables before re-running opam, it's just that any developer using a package system that installs in non-standard locations should *always* have these variables setup. 

Since currently opam has no good story w.r.t. to package system interaction, it seems much better to solve this that way rather than the various attempts made by people mentioned [here](https://github.com/ocaml/opam-repository/issues/6271) which I personally find more obscure than enlighting.

**Comment by AltGr** [comment]
True, opam is currently sensitive to the environment, but advising it may lead people to wrongly believe that `export C_INCLUDE_PATH=... LIBRARY_PATH=...` is a solution, then they'll get bitten later on by the same problem ; in that respect, `opam config setenv` sounds like a good solution to me, it would be more reliable, specific to the switch, and keep working if we sanitise the build environments.

**Comment by dbuenzli** [comment]
> then they'll get bitten later on by the same problem

Precisely not since this should be part of their `.profile`. While this: 

> opam config setenv sounds like a good solution to me, it would be more reliable, specific to the switch

Precisely will, since they will have to do it each time they create a new switch...

But in any case this should be thought out in the more general setting of the `opam`-system package manager interaction.

**Comment by AltGr** [comment]
Oh, I see where the misunderstanding is. I had in mind the case where one has compiled a specific dependency somewhere, and installed in a custom prefix (e.g. `~/src/some-c-project/inst`, or `/opt/foo/`) ; that lib and prefix may only be needed for a specific package and not wanted globally. While it seems you had in mind libraries installed to a more general custom directory (e.g. `~/local`), where indeed the variables should be set in `.profile` at all times.

In any case, we should be careful as to what we advise (`Fix your setup in ~/.profile` vs. `export LIBRARY_PATH`)

**Comment by dbuenzli** [comment]
> In any case, we should be careful as to what we advise (Fix your setup in ~/.profile vs. export LIBRARY_PATH)

Yes ! If can avoid having to tell that to people reporting issues on the opam-repository issue tracker all the time I'd be very grateful.

**Comment by davehayes** [comment]
IMHO you should be really careful advising people to edit shell startup scripts. I can't tell you how many times I've heard "it didn't build for me" from some user only to find that the cause was some obscure environment variable which was inserted to solve another problem some months or years back.

All of the issues I had in my recent opam installs on FreeBSD boiled down to no opam binding support for the standard `/usr/local` prefix which is on most *BSDs.

**Comment by dbuenzli** [comment]
> I can't tell you how many times I've heard "it didn't build for me" from some user only to find that the cause was some obscure environment variable which was inserted to solve another problem some months or years back.

As @ygrek mention, ideally this should be part of the configuration performed by your host platform, the user shouldn't have to perform this.

> All of the issues I had in my recent opam installs on FreeBSD boiled down to no opam binding support for the standard /usr/local prefix which is on most *BSDs.

The problem here is that we are not talking about "`opam` binding support". `opam` is largely oblivious of this. It's how the packages themselves configure, it's a problem with the packages, not opam. Now if you want to chase upstreams so that they fix their packages to detect `/usr/local` be my guest but it doesn't feel like a scalable solution. 

Head to https://github.com/ocaml/opam-repository/issues/3093 for a longer discussion.
